cmd/mt-whisper-importer-reader/conversion.go

```diff
@@ -0,0 +1,209 @@
+		arch := c.archives[i]
+		if arch.SecondsPerPoint > spp {
+			break
+		}
+		smallestArchiveIdx = i
+	}
+
+	return smallestArchiveIdx, largestArchiveIdx
+}
+
+// generates points according to specified parameters by finding and using the best archives as input
+func (c *conversion) getPoints(retIdx int, spp, nop uint32) map[string][]whisper.Point {
+	res := make(map[string][]whisper.Point)
+
+	if len(c.points) == 0 {
+		return res
+	}
+
+	// figure out the range of archives that make sense to use for the requested specs
+	smallestArchiveIdx, largestArchiveIdx := c.findSmallestLargestArchive(spp, nop)
+	rawRes := c.archives[0].SecondsPerPoint
+
+	adjustedPoints := make(map[string]map[uint32]float64)
+	if retIdx > 0 && c.method == "avg" {
+		adjustedPoints["cnt"] = make(map[uint32]float64)
+		adjustedPoints["sum"] = make(map[uint32]float64)
+	} else {
+		adjustedPoints[c.method] = make(map[uint32]float64)
+	}
+
+	// Out of the input archives that we'll use, start with the lowest resolution one by converting
+	// it to the requested resolution and filling the resulting points into adjustedPoints.
+	// Then continue with archives of increasing resolutions while overwriting the generated points
+	// of previous ones.
+	for i := largestArchiveIdx; i >= smallestArchiveIdx; i-- {
+		in := c.points[i]
+		arch := c.archives[i]
+		if arch.SecondsPerPoint == spp {
+			if retIdx == 0 || c.method != "avg" {
+				for _, p := range in {
+					adjustedPoints[c.method][p.Timestamp] = p.Value
+				}
+			} else {
+				for _, p := range in {
+					adjustedPoints["sum"][p.Timestamp] = p.Value
+					adjustedPoints["cnt"][p.Timestamp] = 1
+				}
+			}
+		} else if arch.SecondsPerPoint > spp {
+			method := c.method
+			if method == "avg" && retIdx > 0 {
+				method = "fakeavg"
+			}
+			for m, points := range incResolution(in, method, arch.SecondsPerPoint, spp, rawRes) {
+				for _, p := range points {
+					adjustedPoints[m][p.Timestamp] = p.Value
+				}
+			}
+		} else {
+			methods := []string{c.method}
+			if c.method == "avg" && retIdx > 0 {
+				methods = []string{"sum", "cnt"}
+			}
+			for m, points := range decResolution(in, methods, arch.SecondsPerPoint, spp) {
+				for _, p := range points {
+					adjustedPoints[m][p.Timestamp] = p.Value
+				}
+			}
+		}
+
+	}
+
+	// merge the results that are keyed by timestamp into a slice of points
+	for m, p := range adjustedPoints {
+		for t, v := range p {
+			res[m] = append(res[m], whisper.Point{Timestamp: t, Value: v})
+		}
+		res[m] = sortPoints(res[m])
+	}
+
+	return res
+}
+
+// increase resolution of given points according to defined specs
+func incResolution(points []whisper.Point, method string, inRes, outRes, rawRes uint32) map[string][]whisper.Point {
+	out := make(map[string][]whisper.Point)
+	aggFactor := float64(outRes) / float64(rawRes)
+	for _, inPoint := range points {
+		if inPoint.Timestamp == 0 {
+			continue
+		}
+
+		rangeEnd := inPoint.Timestamp - (inPoint.Timestamp % outRes)
```

Comment:
correct me if i'm wrong:
1) points in whisper are quantized (e.g. for a res of 30s, timestamps in whisper are always divisible by 30 without remainder)
2) thus, all points divide by inRes without remainder
3) outRes < inRes
4) all points divide by outRes without remainder
5) rangeEnd always equals inPoint.Timestamp

Reply:
I'm not sure if i'm misunderstand what you mean, but don't your point 4) and point 5) conflict with each other?
If inRes is 15 and outRes is 10 then some points will not divide by outRes without remainder

Reply:
i updated point 4. had a mistake in there.
i don't have any particular point other than I think 5 is an interesting observation that is probably worth documenting (if i'm correct)

Reply:
added comments

Reply:
your comment says "outRes is > inRes", that's wrong, no?

Reply:
lol right, total confusion